### ISSUE TITLE
DialoGPT: Large-Scale Generative Pre-training for Conversational Response Generation

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#2]
# 논문 정보 (URL, 저자 등)

[http://arxiv.org/abs/1911.00536](http://arxiv.org/abs/1911.00536)

# 개요

- 사람의 대화는 일반적으로 잘알려진 생성 태스크들 (기계 번역, 문서 요약, 파라프레이징) 등에 비해 one-to-many적인 성격이 강하고, informal, noisy, 생략, 문법 에러 등을 포함하는 경우가 많습니다.
- open-domain neural response generation은 아래와 같은 문제점들이 있습니다.
    - content or style inconsistency
    - 장기 문맥 정보의 부재
    - 뻔한 답변 등
- Reddit 코멘트 147M 건을 기반으로 학습한 대화 생성용 GPT입니다.
    - 이 데이터를 학습함으로써, DialoGPT는 대화에서의 P(T,S)의 joint 분포를 더 치밀하게 파악할 수 있게 됩니다.
- 싱글 턴 대화에 있어서는, automatic/human evaluation에서 거의 사람과 동등한 성능을 보였습니다.

# 모델

- GPT-2와 완전히 동일한 구조를 이용합니다.
- 전체 대화 세션의 각 턴을 eos token으로 concat시킵니다.

<img width="311" alt="스크린샷 2020-08-27 오후 4 38 55" src="https://user-images.githubusercontent.com/49383323/91411790-cd3e6680-e883-11ea-8699-0b78bbe02da6.png">

- 모델은 위와 같은 objective로 학습됩니다.
    - 멀티턴 대화일 경우는 아래와 동일한 식이라 할 수 있습니다. (2)
<img width="125" alt="스크린샷 2020-08-27 오후 4 40 09" src="https://user-images.githubusercontent.com/49383323/91411917-fc54d800-e883-11ea-8f42-e86d68edf53c.png">

    - 위 식은 실질적으로 아래 식의 조건부 확률의 곱이라 할 수 있으므로, (3)
<img width="134" alt="스크린샷 2020-08-27 오후 4 39 19" src="https://user-images.githubusercontent.com/49383323/91411818-dcbdaf80-e883-11ea-9ecf-7c7a6f9ef634.png">
    - 결과적으로, (2)식을 옵티마이징한다는 것은 source-target pairs의 모든 (3)식을 옵티마이징한다고 인식할 수 있습니다.

**MMI 리랭킹**

- bland, uninformative 답변을 피하기 위한 전략으로 MMI scoring function을 구현했습니다.
- 우선 top-K sampling을 통해서 여러개의 hypotheses를 생성한 후, backward 모델을 이용하여 P(Source|Hypotheses) 확률을 구하여 리랭킹했습니다.
    - 직관적으로 생각했을 때 뻔한 답변일수록 P(Source|Hypothesis)의 값이 낮을 것 (loss는 크고) 입니다.
- 또한, policy gradient (Williams, 1992)를 이용하여 리워드 R= P(Source|Hypotheses)를 옵티마이즈하는 방법도 시도해보았습니다.
    - 하지만, RL 전략은 hypothesis가 source 문장을 그대로 흉내내는 locally-optimal한 방향으로 수렴하는 것을 발견했는데, 이는 RNN 구조와 달리 Transformer는 representation power가 너무 강력하기 때문이라고 추측하고 있습니다.
    - 이 부분을 더 정확히 분석하고 발전시키는 것이 Future work으로 언급되어 있습니다.

# 데이터

- Reddit의 2005년~2017년 댓글을 이용합니다.
    - Reddit discussion은 기본적으로 tree 구조를 띄고 있기 때문에, root node (첫 댓글)에서 leaf node로 가는 모든 경로를 추출하여 각 경로를 하나의 멀티턴 training instance로 간주했습니다.
- 아래와 같은 필터링 과정을 거쳤습니다.
    1. Source 혹은 Target에 URL을 포함하는 인스턴스를 제거
    2. target에 3 단어 이상 단어가 반복되는 인스턴스를 제거
    3. response에 빈도 상위 50개 영어 단어가 하나도 포함이 안된 인스턴스를 제거
        - the, of, a 등의 단어가 하나도 없는 경우는 영어 문장이 아닐 경우가 높다고 판단
    4. response에 "[", "]" 가 포함되어 있는 경우 markup language로 판단하여 제거
    5. source와 target이 둘다 200단어를 넘길 경우 제거
    6. target에 offensive language가 포함되어 있을 경우 제거
        - phrase 매칭으로 판단했다고 함
        - offensive content를 포함할만한 서브레딧들도 모두 제거
    7. 전체 데이터 상에서 1000번이상 등장한 tri-gram이 response에 90%이상을 차지하는 bland한 답변을 제거
        - 이러한 답변들은 전체 데이터에서 1% 이상을 차지했다고 함.

# 실험

**실험 상세 사항**

- 3가지 사이즈의 모델로 학습을 해봤습니다.

<img width="343" alt="스크린샷 2020-08-27 오후 4 40 43" src="https://user-images.githubusercontent.com/49383323/91411967-0d9de480-e884-11ea-85e3-b09b723a4d23.png">

- vocab은 50,257개를 썼고, V100 16개로 학습했습니다.
- Noam lr sceduler with 16000 warm-up steps을 이용했고, lr은 validation loss를 베이스로 정했습니다.
- 각 모델은 validation loss에 더이상 변화가 없을 때까지 학습을 진행했고, 그 결과 small, medium은 각각 5에폭, large는 3에폭까지 반복했습니다.

**학습 최적화**

- 학습을 빠르게 진행하기 위해, 모든 학습 데이터를 lazy-loading database file로 압축하여 데이터가 필요할 때만 로딩될 수 있도록 (액세스 빈도를 줄이기 위해 큰 청크를 프리패치) 했습니다.
    - TF Record를 썼다는 소리인듯
- 학습을 스케일하기 위해 비동기 데이터 프로세스를 활용했습니다.
- 이 결과, 학습 시간이 GPU 갯수에 따라 선형적으로 감소할 수 있었고, 추가적으로 학습 처리량을 늘리기 위해 같은 배치를 비슷한 길이의 대화들로 구성하기 위해 dynamic batching 전략을 이용했습니다.

**실험 결과**

<img width="690" alt="스크린샷 2020-08-27 오후 4 41 52" src="https://user-images.githubusercontent.com/49383323/91412081-38883880-e884-11ea-85f2-6b572583323d.png">

- OpenAI GPT-2를 파인튜닝한 것과 scratch부터 학습한 모델을 비교해본 결과
    - 작은 모델의 경우는 OpenAI GPT-2를 파인튜닝하는 경우가 performance gain이 크고, 모델이 커질수록 성능 차이가 적어지는 것을 알 수 있었습니다.
    - 그리고 모델의 사이즈가 커짐에 따라 모델의 성능이 좋아지는 것을 알 수 있었습니다.
- MMI 리랭킹
    - top-10 sampling으로 16개의 샘플을 만든뒤, backward model이 제일 낮은 로스를 보이는 답변을 최종 답변으로 골랐습니다.
    - MMI 리랭킹은 NIST, METEOR,DIST 등의 값이 더 커지고, BLEU의 값은 살짝 낮아지는데, 이는 더 Diverse한 답변이 나온다는 것을 의미합니다.
- Human Evaluation 결과 충분히 인간에 근접한 성능이 보고되었습니다.
    - Relevance, Informative, Human-likeness를 평가했고, System A, System B중 어느쪽이 더 우수한가?를 기준으로 레이블링을 진행했습니다.
    - 심지어 MMI DialoGPT는 human performance를 뛰어넘었는데, 이는 사람의 답변이 불규칙적이고, 독특한 특징이 있는 경우가 있고, 인터넷 밈 등을 포함하는 답변이 더러 포함되어있기 때문이라고 추측합니다.

<img width="717" alt="스크린샷 2020-08-27 오후 4 42 18" src="https://user-images.githubusercontent.com/49383323/91412118-48a01800-e884-11ea-89c8-4b5c02d46ab3.png">

**모델의 한계**

- 모델은 offensive, gender-bias 등이 담긴 문장을 생성할 수도 있습니다.

# 논의할만한 내용

- 여기서도 OpenAI GPT를 파인튜닝하는 방식이 꽤 괜찮은 성능이라고 하는데, SKT KoGPT 파인튜닝하는 방식도 한번 시도해봄직한듯.

# 관련된 논문

- Icecaps (Shiv et al., 2019): 답변 생성 툴킷
